torch/csrc/api/src/nn/modules/activation.cpp

```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
   stream << ")";
 }
 
+// ============================================================================
+
+MultiheadAttentionImpl::MultiheadAttentionImpl(const MultiheadAttentionOptions& options_)
+    : options(options_) {
+  reset();
+}
+
+std::tuple<Tensor, Tensor> MultiheadAttentionImpl::forward(
+  const Tensor& query, const Tensor& key,
+  const Tensor& value, const Tensor& key_padding_mask,
+  bool need_weights, const Tensor& attn_mask) {
+  if (_qkv_same_embed_dim) {
+    return F::multi_head_attention_forward(
+      query, key, value, 
+      F::MultiheadAttentionForwardOptions(
+        /*embed_dim_to_check=*/options.embed_dim(),
+        /*num_heads=*/options.num_heads(),
+        /*in_proj_weight=*/in_proj_weight,
+        /*in_proj_bias=*/in_proj_bias,
+        /*bias_k=*/bias_k,
+        /*bias_v=*/bias_v,
+        /*add_zero_attn=*/options.add_zero_attn(),
+        /*dropout_p=*/options.dropout(),
+        /*out_proj_weight=*/out_proj->weight,
+        /*out_proj_bias=*/out_proj->bias
+      ).training(is_training())
+       .key_padding_mask(key_padding_mask)
+       .need_weights(need_weights)
+       .attn_mask(attn_mask)
+    );
+  } else {
+    return F::multi_head_attention_forward(
+      query, key, value, 
+      F::MultiheadAttentionForwardOptions(
+        /*embed_dim_to_check=*/options.embed_dim(),
+        /*num_heads=*/options.num_heads(),
+        /*in_proj_weight=*/in_proj_weight,
+        /*in_proj_bias=*/in_proj_bias,
+        /*bias_k=*/bias_k,
+        /*bias_v=*/bias_v,
+        /*add_zero_attn=*/options.add_zero_attn(),
+        /*dropout_p=*/options.dropout(),
+        /*out_proj_weight=*/out_proj->weight,
+        /*out_proj_bias=*/out_proj->bias
+      ).training(is_training())
+       .key_padding_mask(key_padding_mask)
+       .need_weights(need_weights)
+       .attn_mask(attn_mask)
+       .use_separate_proj_weight(true)
+       .q_proj_weight(q_proj_weight)
+       .k_proj_weight(k_proj_weight)
+       .v_proj_weight(v_proj_weight)
+    );
+  }
+}
+
+void MultiheadAttentionImpl::reset() {
+  _qkv_same_embed_dim = options.kdim() == options.embed_dim() &&
+                        options.vdim() == options.embed_dim();
+  head_dim = options.embed_dim() / options.num_heads();
+  TORCH_CHECK(head_dim * options.num_heads() == options.embed_dim(),
+              "embed_dim must be divisible by num_heads");
+  _reset_parameters();
+}
+
+void MultiheadAttentionImpl::_reset_parameters() {
+  using namespace torch::nn::init;
+  if (!_qkv_same_embed_dim) {
+    q_proj_weight = torch::empty({options.embed_dim(), options.embed_dim()});
+    k_proj_weight = torch::empty({options.embed_dim(), *options.kdim()});
+    v_proj_weight = torch::empty({options.embed_dim(), *options.vdim()});
```

Comment:
we probably need to add the equivalent of `self.register_parameter('in_proj_weight', None)` here, to mirror the Python version better